src/audio/dai-zephyr.c

```diff
@@ -487,6 +487,20 @@ static int dai_playback_params(struct comp_dev *dev, uint32_t period_bytes,
 
 		comp_info(dev, "dai_playback_params() fifo 0x%x", fifo);
 
+		err = dma_get_attribute(dd->dma->z_dev, DMA_ATTR_MAX_BLOCK_COUNT,
+					&max_block_count);
+		if (err < 0) {
+			comp_err(dev, "dai_playback_params(): could not get dma attr max block count, err = %d", err);
+			goto out;
+		}
+
+		if (max_block_count < period_count) {
+			comp_info(dev, "dai_playback_params(): block count = %d not supported by DMA", period_count);
+			buf_size = period_count * period_bytes;
+			period_count = max_block_count;
+			period_bytes = buf_size / period_count;
```

Comment:
Maybe add an assert for the case where this does not divide evenly? Should never happen of course.

Reply:
hmm, maybe there's some implicit rule that I'm not aware of, but by just looking at this, it might well not divide evenly. We enter here when the caller requested more periods than the DMA can handle. We then calculate a buffer size from that larger block count and divide it by the the DMA maximum block count. So if max was 16 and the caller requested 17 and the period size isn't a multiple of 16, this won't divide well. Or is this meant for cases where the requested period count is a multiple of the block count, fitting multiple periods in a single DMA block?

Reply:
The only one real use case is limiting period_count to 1 for HDA DAI so more like as you said - fitting multiple period in a single DMA block, I thought we can make this more generic.
Division problem we can quickly solve with assert or ALIGN_DOWN but Im now not sure if this will result in a valid config, lets say 4*128 is still valid if requested 8*64?

Reply:
Updated with alignment check and period count decreasing, please take a look if it's not over complicated